### ISSUE TITLE
Add Emoji Support

### ISSUE THREAD

**Issue by omergoktas** [rkkr/simple-keyboard#438]
I love this project, and it's my main keyboard, except I often feel the need to be able to type emojis. I wish there was built-in emoji support so we didn't have to open WhatsApp or Telegram just to copy emoji symbols.

**Comment by Stardisk** [comment]
You know, I have chosen this keyboard because it **doesn't contain** emojis and other useless stuff. 

So, dear author, please don't listen to him and don't add emojis, and thank you for this great keyboard

**Comment by omergoktas** [comment]
A simple switch in the settings could easily solve your problem (enable/disable emojis).

"dEaR aUtHor dOnT lİsTeN tO hiM bEcAuSe I lİkE iT tHiS wAy"

How old are you, friend?

🤦‍♂️

**Comment by Stardisk** [comment]
I'm 31 years old, and what?
You can use default GBoard or any other keyboard with emojis, gifs, stickers, autocompleting and other garbage, which make **key**board more heavy and slow. By the way, your PC keyboard (if you use PC) doesn't have all these stuff, but you don't write issues to your keyboard manufacturer and require to add emojis or something else. And i don't need these things on my Android smartphone, so I've chosen Simple Keyboard because it lightweight, fast and simple, like PC keyboard.

**Comment by hueldoeu** [comment]
thank you for simple keyboard. on my lg g5 i had a great keyboard. now on my huawei p20 pro i moved from ms swiftkey to gboard to celia (maybe from huawei). gboard was nice but the enter key had some extra features who were unneccessary and interrupted working. celia keyboard had very bad issues, while the voice to text software worked very well, the typing of links in a browser alway had a space when i typed 50 % of the domain and pressed enter.

**Comment by hueldoeu** [comment]
sadly no autocorrect

**Comment by Cwpute** [comment]
I use a lot of emojis in my day to day casual communication, as well as in some official communications on social networks, and this would be a very welcome feature indeed. But i do understand that this feature is not desired by the maintainer of this app so it's safe to assume it will never be implemented.

Although i'd like to know the reasons that motivate them not to have emojis as an option ?
This keyboard being so small, would emojis inflate its size so much ? I'm genuinely asking.

**Comment by hueldoeu** [comment]
> I use a lot of emojis in my day to day casual communication, as well as in some official communications on social networks, and this would be a very welcome feature indeed. But i do understand that this feature is not desired by the maintainer of this app so it's safe to assume it will never be implemented.
> 
> Although i'd like to know the reasons that motivate them not to have emojis as an option ? This keyboard being so small, would emojis inflate its size so much ? I'm genuinely asking.

emojis started as :-) :-( :-P :-/ :-* etc. and i actually prefer them more than e. g. windows-key + . (dot) 👨🏻👱🏻‍♀️💂🏻‍♂️👨🏻‍💻😀🙂😥😝😨🤥🐱‍💻🐭🐐🦕🐛

however i really need a huge clipboard manager. on windows i work with ditto-cp a lot https://ditto-cp.sourceforge.io/ . i wish it will be add one day.

**Comment by herrbrixter** [comment]
Just install Heliboard. [https://github.com/Helium314/HeliBoard](https://github.com/Helium314/HeliBoard)

I uninstalled Simple Keyboard because I can't use it without emojis. Heliboard has emoji support.

**Comment by herrbrixter** [comment]
> Heliumboard

It's Heliboard, not Heliumboard. [https://github.com/Helium314/HeliBoard](https://github.com/Helium314/HeliBoard)

**Comment by herrbrixter** [comment]
> By the way, your PC keyboard (if you use PC) doesn't have all these stuff, but you don't write issues to your keyboard manufacturer and require to add emojis or something else.

Most keyboard apps have emoji support so people are asking for it. No PC keyboard as far as I know has emoji support so nobody asks for it. Also most people use emojis on messaging apps, not in emails.

**Comment by makerandreas** [comment]
The concept of this keyboard is actually pretty good. It's light, and there aren't many weird things about it. I'd also love to see an emoji feature on this keyboard. After seeing that some people liked it and others didn't, I think it'd be fair to add the emoji feature but include a setting so you can enable or disable it as you like. If you don't want to use emojis, you can just disable the feature.

**Comment by user44000** [comment]
Maybe an extension could be created for this keyboard that could add emoji support and other features people might want that won't be added to this keyboard.

**Comment by Cwpute** [comment]
At this point, the best solution would probably be for someone to fork Simple Keyboard and add emoji support.

**Comment by maagenta** [comment]
Okay guys, I made it!

![image](https://github.com/user-attachments/assets/9171db0a-d9ac-4722-856e-bdc98fb4a17d)

The code is here:
[https://github.com/maagenta/simple-keyboard/tree/emoji-feature](https://github.com/maagenta/simple-keyboard/tree/emoji-feature)

The APK size is 2.2MB and the keyboard uses 3.2MB when installed.

Also, in settings there is the option "Show Emoji Key" to enable/disable the feature.

<img src="https://github.com/user-attachments/assets/a9469922-f4d9-4b43-ade3-39f623c8e722" width="200px">

I'll make a pull request from my fork. If they deny my request, I will build the distributable packages in my repo.

**Comment by rkkr** [comment]
@maagenta you don't need to bother making a pull request. With an AndroidX dependency that more than doubles the APK size it's going to be denied.

**Comment by maagenta** [comment]
> [@maagenta](https://github.com/maagenta) you don't need to bother making a pull request. With an AndroidX dependency that more than doubles the APK size it's going to be denied.

The AndroidX dependency I was using was `androidx.recyclerview:recyclerview:1.3.2`
                                                                                                                                                        
I removed it and replaced it with GridView instead.                                                                                                   
                                                                                                                                                        
OMG, the APK is now 807KB!                                                                                                                            
                  
I think the signed APK could be even smaller (around 700KB).                                                                                          
   
The app size is 823KB and works perfectly!

What do you think now ?

**Comment by rkkr** [comment]
...and now you see why all the apps are in 100s of MBs. And no amount of optimizations, minifications and bundles can compensate the bloat of google.

The goals of this app probably still stand. You're always free to publish the APKs and distribute thru Play Store and others if you want to. I think I've already seen a few emoji forks published.

If I were to add emoji support I'd diff this to LatinIME. This is a LatinIME fork with emoji functionality removed so you could just "unremove it". And the emoji key along the space will cause issues for layouts that put ZWJ and ZWNJ charaters there.

**Comment by maagenta** [comment]
> ...and now you see why all the apps are in 100s of MBs. And no amount of optimizations, minifications and bundles can compensate the bloat of google.
> 
> The goals of this app probably still stand. You're always free to publish the APKs and distribute thru Play Store and others if you want to. I think I've already seen a few emoji forks published.
> 
> If I were to add emoji support I'd diff this to LatinIME. This is a LatinIME fork with emoji functionality removed so you could just "unremove it". And the emoji key along the space will cause issues for layouts that put ZWJ and ZWNJ charaters there.

I really appreciate your answer!

Now the app is 652KB having 3200 emojis available. Your suggestion of removing androidx dependencies really helped me. I also decided to code the emoji rendering by calling the GPU directly and caching each emoji in RAM, so the GPU can read them immediately without using any CPU resources. It makes the keyboard use a lot of RAM (140MB) in exchange for practically no CPU usage. The CPU uses much more energy than RAM, and my phone has 8GB of RAM, so for me that is completely acceptable.

I understand your point about not adding emojis. Rendering 3200 full color characters will always have performance consequences. You can make the keyboard avoid using RAM by using lazy-loading, but it will make the CPU work more, so it's nice to have a simple keyboard that avoids this kind of problem.

Also, thanks for the suggestion of cloning LatinIME. I tried it but I think my emoji implementation works better for me than LatinIME: better performance, all emojis from EmojiV15 (3200) available, and simplicity.

It doesn't have any problems with layouts using ZWJ and ZWNJ. I handled those cases with switches in the XML layout.

**Comment by maagenta** [comment]
Ok, here is the fork with emoji support (the apk is 703KB):
[https://github.com/maagenta/kokey](https://github.com/maagenta/kokey)

**Comment by user44000** [comment]
I don't think this is getting added so I think the issue should be closed. There's no shortage of keyboard apps that have emoji support so this one doesn't need it. If people want emojis they can use a different keyboard.

**Comment by slvlng** [comment]
> The goals of this app probably still stand.

It's great that the developer maintained an independent vision while creating this minimalist-style keyboard, continuing to improve it.

So, someone even forked the app and added emoji support. And it's funny that almost none of those who requested such a seemingly essential feature show any interest in it...

**Comment by omergoktas** [comment]
> And it's funny that almost none of those who requested such a seemingly essential feature show any interest in it...

Right? It's as if they moved on with life and switched to iPhone or something... sigh

**Comment by Cwpute** [comment]
fantastic job @maagenta !! thanks
